File: java/P13.java
// P13 Run-length encoding of a list (direct solution).

import core.List;

import static core.List.*;
import static core.Pair.*;

class P13 {
    // Imperative, list of objects
    @SuppressWarnings("unchecked")
    List f1(List list) {
        if (list.isEmpty()) return nil();

        List acc = nil();
        Object x = list.head();
        int n = 1;

        for (Object each : list.tail()) {
            if (each.equals(x)) {
                n++;
            } else {
                acc = cons(makeValue(n, x), acc);
                x = each;
                n = 1;
            }
        }

        acc = cons(makeValue(n, x), acc);

        return acc.reverse();
    }

    // Factory method
    private Object makeValue(int n, Object x) {
        return n == 1 ? x : pair(n, x);
    }
}

File: clojure/P13.clj
; P13 Run-length encoding of a list (direct solution).

; Factory function
(defn encode [n x]
  (if (= 1 n) x [n x]))

; Using "split-with", so there's still some kind of sublists
(defn f1 [l]
  (if (seq l)
    (let [h (first l) t (rest l) [a b] (split-with #(= % h) t)]
      (cons (encode (inc (count a)) h) (f1 b)))
    '()))

; Recursion, "look-behind"
(defn f2 [l]
  (defn f [n x xs]
    (if (seq xs)
      (let [y (first xs) ys (rest xs)]
        (if (= y x)
          (f (inc n) y ys)
          (cons (encode n x) (f 1 y ys))))
      (list (encode n x))))
  (if (seq l) (f 1 (first l) (rest l)) '()))